hudi-common/src/main/java/org/apache/hudi/common/table/read/FileGroupRecordBufferInitializer.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.table.read;
+
+import org.apache.hudi.common.config.HoodieMemoryConfig;
+import org.apache.hudi.common.config.HoodieReaderConfig;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.engine.HoodieReaderContext;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.table.PartialUpdateMode;
+import org.apache.hudi.common.table.log.HoodieMergedLogRecordReader;
+import org.apache.hudi.common.util.ConfigUtils;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.storage.HoodieStorage;
+
+import java.io.Closeable;
+import java.util.List;
+
+import static org.apache.hudi.common.util.ConfigUtils.getIntWithAltKeys;
+
+/**
+ * This interface defines the contract for initializing a {@link FileGroupRecordBuffer} for a given file group.
+ * The implementation is expected to establish the record buffer and populate it with the records from the log files.
+ * @param <T> the engine specific record type
+ */
+public interface FileGroupRecordBufferInitializer<T> {
```

Comment:
This is the largest change of the PR. The initializer is responsible for creating the record buffer and loading it with any records. This may be useful for the MergeHandle changes as well.

Also I am very open to any suggestions on naming, I truly dislike this part of coding 😅

Reply:
Should this be refactored to be a single factory class so `getRecordBuffer` or the create method can take `reuse` as another parameter to decide the record buffer to construct and return, to simplify the initialization, instead of having both layers of `DefaultFileGroupRecordBufferInitializer` and `ReusableFileGroupRecordBufferInitializer`, vs. `FileGroupRecordBuffer` and `ReusableKeyBasedRecordBuffer`?